hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMapProxy.java

```diff
@@ -1036,14 +1037,50 @@ public boolean isEmpty() {
 
     @Override
     public void putAll(Map<? extends K, ? extends V> m) {
-        MapEntrySet entrySet = new MapEntrySet();
+        ClientPartitionService partitionService = getContext().getPartitionService();
+        int partitionCount = partitionService.getPartitionCount();
+        List<Future<?>> futures = new ArrayList<Future<?>>(partitionCount);
+        MapEntrySet[] entrySetPerPartition = new MapEntrySet[partitionCount];
+        
         for (Entry<? extends K, ? extends V> entry : m.entrySet()) {
+            checkNotNull(entry.getKey(), NULL_KEY_IS_NOT_ALLOWED);
+            checkNotNull(entry.getValue(), NULL_VALUE_IS_NOT_ALLOWED);
+            
             final Data keyData = toData(entry.getKey());
             invalidateNearCache(keyData);
+            
+            int partitionId = partitionService.getPartitionId(entry.getKey());
+            MapEntrySet entrySet = entrySetPerPartition[partitionId];
+            if (entrySet == null) {
+                entrySet = new MapEntrySet();
+                entrySetPerPartition[partitionId] = entrySet;
+            }
+            
             entrySet.add(new AbstractMap.SimpleImmutableEntry<Data, Data>(keyData, toData(entry.getValue())));
         }
-        MapPutAllRequest request = new MapPutAllRequest(name, entrySet);
-        invoke(request);
+        
+        for (int partitionId = 0; partitionId < entrySetPerPartition.length; partitionId++) {
+            MapEntrySet entrySet = entrySetPerPartition[partitionId];
+            if (entrySet != null) {
+                //If there is only one entry, consider how we can use MapPutRequest without
+                //having to get back the return value.
+                MapPutAllRequest request = new MapPutAllRequest(name, entrySet);
+                
+                //invoke by partition owner address rather than id, as
+                //the latter will embed the partitionId into the Packet payload,
+                //routing the invocation to the operation service, which does 
+                //not handle MapPutAllRequest and other similar requests
+                futures.add(new ClientInvocation(getClient(), request, partitionService.getPartitionOwner(partitionId)).invoke());
```

Comment:
This does not look right. We already gather the keys that should go to one partition. We should not make use of a multi partition request here. 

The problem with this approach there will be no check on server side if this requests is running in correct node. Meaning of correct here is the node that owns of the partition. And partition table can change anytime.

We can modify MapPutAllRequest so that it will invoke an operation that will run on one node. It will be PartitionClientRequest instead of AllPartitionRequest. Also we need to the similar for MessageTasks(new client)